python/shotgun_fields/bubble_widget.py

```diff
@@ -0,0 +1,495 @@
+
+        The ``bubble_id`` should correspond to the unique ID returned by the
+        :meth:`.add_bubble` method.
+
+        :param bubble_id: The id of the bubble to retrieve.
+        :return: A bubble widget or ``None`` if not match is found
+        :rtype: :class:`.BubbleWidget`
+        """
+
+        if not bubble_id in self._formats:
+            return None
+
+        text = self.toPlainText()
+        cursor = self.textCursor()
+
+        # loop over each character until a replacement character with a known
+        # format that matches the supplied id is found.
+        for i in range(0, len(text)):
+            if not text[i] == self._OBJECT_REPLACEMENT_CHAR:
+                continue
+
+            cursor.setPosition(i+1, QtGui.QTextCursor.MoveAnchor)
+            format = cursor.charFormat()
+
+            if not self._formats[bubble_id] == format:
+                continue
+
+            # bubble is in the text
+            return self._bubble_text_object.get_bubble(bubble_id)
+
+        return None
+
+    def get_bubbles(self):
+        """
+        Similar to ``get_bubble``, but returns all bubble widgets.
+
+        :return: List of :class:`.BubbleWidget` classes
+        :rtype: list
+        """
+
+        text = self.toPlainText()
+        cursor = self.textCursor()
+
+        formats = []
+        bubbles = []
+
+        # find formats for bubbles in the text editor
+        for i in range(0, len(text)):
+            if not text[i] == self._OBJECT_REPLACEMENT_CHAR:
+                continue
+
+            cursor.setPosition(i+1, QtGui.QTextCursor.MoveAnchor)
+            format = cursor.charFormat()
+
+            if not format in self._formats.values():
+                continue
+
+            formats.append(format)
+
+        # get the corresponding bubbles for each found format
+        for bubble_id in self._formats:
+            if self._formats[bubble_id] in formats:
+                bubbles.append(self.get_bubble(bubble_id))
+
+        return bubbles
+
+    def get_typed_text(self):
+        """
+        Returns a :obj:`str` representing the text typed in the editor.
+        """
+
+        char_list = [c for c in self.toPlainText() if c != self._OBJECT_REPLACEMENT_CHAR]
+        return "".join(char_list)
+
+    def remove_bubble(self, bubble_id):
+        """
+        Remove a bubble matching the supplied id.
+
+        :param int bubble_id: The unique id of the bubble to reomve.
+        :return: The removed qt widget
+
+        The ``bubble_id`` should correspond to the unique ID returned by the
+        :meth:`.add_bubble` method.
+        """
+
+        if not self.get_bubble(bubble_id):
+            return
+
+        text = self.toPlainText()
+        cursor = self.textCursor()
+
+        # locate the bubble by iterating over each character in the editor.
+        # when the match is found, remove it.
+        for i in range(0, len(text)):
+            if not text[i] == self._OBJECT_REPLACEMENT_CHAR:
```

Comment:
good call. found a bunch of others in here too. i think i did it this way once and cut/pasted a bunch.